blocktopmenu.php

```diff
@@ -526,50 +574,87 @@ protected function makeMenu()
                     $supplier = new Supplier((int)$id, (int)$id_lang);
                     if (!is_null($supplier->id)) {
                         $link = new Link;
-                        $this->_menu .= '<li'.$selected.'><a href="'.Tools::HtmlEntitiesUTF8($link->getSupplierLink((int)$id, $supplier->link_rewrite)).'" title="'.$supplier->name.'">'.$supplier->name.'</a></li>'.PHP_EOL;
+                        $supplierNode = $this->createNode(BlocktopmenuNode::TYPE_MANUFACTURER, array(
+                            'id' => $id,
+                            'link' => Tools::HtmlEntitiesUTF8($link->getSupplierLink((int)$id, $supplier->link_rewrite)),
+                            'name' => Tools::safeOutput($supplier->name),
+                        ));
+                        $items[] = $supplierNode;
+                        $this->_menu .= self::renderNode($supplierNode);
                     }
                     break;
 
                 case 'SHOP':
-                    $selected = ($this->page_name == 'index' && ($this->context->shop->id == $id)) ? ' class="sfHover"' : '';
+                    $selected = $this->page_name == 'index' && $this->context->shop->id == $id;
                     $shop = new Shop((int)$id);
                     if (Validate::isLoadedObject($shop)) {
                         $link = new Link;
-                        $this->_menu .= '<li'.$selected.'><a href="'.Tools::HtmlEntitiesUTF8($shop->getBaseURL()).'" title="'.$shop->name.'">'.$shop->name.'</a></li>'.PHP_EOL;
+                        $shopNode = $this->createNode(BlocktopmenuNode::TYPE_SHOP, array(
+                            'id' => $id,
+                            'link' => Tools::HtmlEntitiesUTF8($shop->getBaseURL()),
+                            'name' => Tools::safeOutput($shop->name),
+                        ));
+                        $shopNode->setSelected($selected);
+                        $items[] = $shopNode;
+                        $this->_menu .= self::renderNode($shopNode);
                     }
                     break;
                 case 'LNK':
                     $link = MenuTopLinks::get((int)$id, (int)$id_lang, (int)$id_shop);
-                    if (count($link)) {
+                    if (count($link) == 1) {
                         if (!isset($link[0]['label']) || ($link[0]['label'] == '')) {
                             $default_language = Configuration::get('PS_LANG_DEFAULT');
                             $link = MenuTopLinks::get($link[0]['id_linksmenutop'], $default_language, (int)Shop::getContextShopID());
                         }
-                        $this->_menu .= '<li><a href="'.Tools::HtmlEntitiesUTF8($link[0]['link']).'"'.(($link[0]['new_window']) ? ' onclick="return !window.open(this.href);"': '').' title="'.Tools::safeOutput($link[0]['label']).'">'.Tools::safeOutput($link[0]['label']).'</a></li>'.PHP_EOL;
+                        $linkNode = $this->createNode(BlocktopmenuNode::TYPE_LINK, array(
```

Comment:
While it's a lot nice now, I fear that direct form field-HTML attribute may not match future need of extensibility.
For the sake of feature-equivalence the hunk seems stuck to the HTML attribute supported up to now.
But the blocktopmenu could have a lot of other uses. I currently have a link using ``target="_blank"`` and another triggering a javascript function. These are still not possible with the new implementation.
(I know we are stuck to simplicity because the current table-schema isn't indefinitely extensible)